folly/detail/Iterators.h

```diff
@@ -93,9 +93,10 @@ class IteratorFacade {
    * this and the `operator==(D const&) const` method above.
    */
 
-  template <class D2>
-  typename std::enable_if<std::is_convertible<D, D2>::value, bool>::type
-  operator==(D2 const& rhs) const {
+  template <class D2,
+    typename std::enable_if_t<!std::is_same<D, D2>::value, int> = 0,
+    typename std::enable_if_t<std::is_convertible<D, D2>::value, int> = 0>
```

Comment:
These two `typename`s don't seem necessary.

Reply:
I think they are in C++ 20. Let's take an example:
```
  854 |   return find(key) != items().end() ? 1u : 0u;                                                                                                                                                                                                                               |          ~~~~~~~~~ ^~ ~~~~~~~~~~~~~
      |              |                   |
      |              |                   folly::dynamic::const_item_iterator
      |              folly::dynamic::const_item_iterator
```
In C++20, evaluating `operator!=` will lead to add `operator!=`, `operator==` and reversed `operator==` in the candidates list. If you don't desambiguate `operator==` when the type is the same you will have twice the same primary operator.

This said, these two lines simply implement the comment just above:
```
   * Does a conversion of D (or D reference) to D2, if one exists (otherwise
   * this is disabled).  Disabled if D and D2 are the same, to disambiguate
   * this and the `operator==(D const&) const` method above.
```
Regards,
Laurent

Reply:
The overload control looks necessary but the use of the keyword `typename` seems superfluous.

Reply:
True. PR updated.